cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -676,7 +677,351 @@ void AddPrimitiveIfElseKernels(const std::shared_ptr<ScalarFunction>& scalar_fun
+    const int32_t width =
+        checked_cast<const FixedSizeBinaryType&>(*values.type).byte_width();
+    uint8_t* next = out_values + (width * offset);
+    const auto& scalar = checked_cast<const ScalarType&>(values);
+    const auto value = scalar.value.ToBytes();
+    for (int i = 0; i < length; i++) {
+      std::memcpy(next, value.data(), width);
+      next += width;
+    }
+  }
+  static void CopyArray(const ArrayData& array, uint8_t* out_values, const int64_t offset,
+                        const int64_t length) {
+    const int32_t width =
+        checked_cast<const FixedSizeBinaryType&>(*array.type).byte_width();
+    uint8_t* next = out_values + (width * offset);
+    const auto* in_values = array.GetValues<uint8_t>(1, (offset + array.offset) * width);
+    std::memcpy(next, in_values, length * width);
+  }
+};
+// Copy fixed-width values from a scalar/array datum into an output values buffer
+template <typename Type>
+void CopyValues(const Datum& values, uint8_t* out_valid, uint8_t* out_values,
+                const int64_t offset, const int64_t length) {
+  using Copier = CopyFixedWidth<Type>;
+  if (values.is_scalar()) {
+    const auto& scalar = *values.scalar();
+    if (out_valid) {
+      BitUtil::SetBitsTo(out_valid, offset, length, scalar.is_valid);
+    }
+    Copier::CopyScalar(scalar, out_values, offset, length);
+  } else {
+    const ArrayData& array = *values.array();
+    if (out_valid) {
+      if (array.MayHaveNulls()) {
+        arrow::internal::CopyBitmap(array.buffers[0]->data(), array.offset + offset,
+                                    length, out_valid, offset);
+      } else {
+        BitUtil::SetBitsTo(out_valid, offset, length, true);
+      }
+    }
+    Copier::CopyArray(array, out_values, offset, length);
+  }
+}
+
+struct CaseWhenFunction : ScalarFunction {
+  using ScalarFunction::ScalarFunction;
+
+  Result<const Kernel*> DispatchBest(std::vector<ValueDescr>* values) const override {
+    RETURN_NOT_OK(CheckArity(*values));
+    std::vector<ValueDescr> value_types;
+    for (size_t i = 0; i < values->size() - 1; i += 2) {
+      ValueDescr* cond = &(*values)[i];
+      if (cond->type->id() == Type::NA) {
+        cond->type = boolean();
+      }
+      if (cond->type->id() != Type::BOOL) {
+        return Status::Invalid("Condition arguments must be boolean, but argument ", i,
+                               " was ", cond->type->ToString());
+      }
+      value_types.push_back((*values)[i + 1]);
+    }
+    if (values->size() % 2 != 0) {
+      // Have an ELSE clause
+      value_types.push_back(values->back());
+    }
+    EnsureDictionaryDecoded(&value_types);
+    if (auto type = CommonNumeric(value_types)) {
+      ReplaceTypes(type, &value_types);
+    }
+
+    const DataType& common_values_type = *value_types.front().type;
+    auto next_type = value_types.cbegin();
+    for (size_t i = 0; i < values->size(); i += 2) {
+      if (!common_values_type.Equals(next_type->type)) {
```

Comment:
Given the definition of `ReplaceTypes`, this check doesn't seem necessary?
cc @bkietz  for a second opinion.

Reply:
It seems to me we could restructure "case_when" and yield a less daunting interface:

```python
case(
  when(cond_0, value_0),
  when(cond_1, value_1),
  value_else
)
```

Where `when` masks slots with null wherever its condition is not true and `case` is only a variadic coalescing function (takes the first non null).

We'd be allocating a new null bitmap on each call to `when` which is not ideal. However typing is far clearer (`when(cond: Boolean, value: T): T`, `case(...values: T): T`) and case/when can be independently unit tested.

@pitrou  what do you think?

Reply:
Note: variadic coalesce is independently useful https://issues.apache.org/jira/browse/ARROW-13136

Reply:
Another way to transpose this would be:

```python
case(
  first_true_in(cond_0, cond1),
  value0,
  value1,
  value_else
)
```

Where `first_true_in` returns the index of the first `true` and `case`'s first argument is an index into the rest of its arguments:

```
assert first_true_in([True, False, False], [True, True, False]) == [0, 1, None]
assert case([0, 1, None], 'a b c'.split(), 'd e f'.split(), 'x y z'.split()) == 'a e z'.split()
```

Reply:
In the second example, case would essentially be 'choose' (ARROW-13220). I'm curious which would be faster; the coalesce PR is implemented columnwise (for fixed-width types) and tries to do block copies, while the 'choose' PR I'm looking at is necessarily rowwise. Probably depends on the distribution of the inputs.

Might the masking 'when' kernel and 'first_true_in' be useful kernels to have anyways?

Reply:
I think the masking 'when' and 'first_true_in' are definitely independently useful

Reply:
I could get choose/coalesce in first, then put up candidate PRs for case-when to compare the implementations.

Reply:
SGTM, thanks!

Reply:
I think that 

```
case(
  when(cond_0, value_0),
  when(cond_1, value_1),
  value_else
)
```

is not necessarily identical to `case_when(cond_0, value_0, cond_1, value_1, value_else)` as done in this PR, when it comes to nulls in the values? 

For example, if you have a null in `value_0`, it should be chosen if the corresponding bit of `cond_0` is True. However, if you implement that with a masking `when` and then coalesce (taking the first non null), it would not pick the null value.

Reply:
Hmm, you're right…

We can still try the 'choose' based version though I would expect that to be slower since it works rowwise.

Reply:
I put up PRs for both choose: #10642 and coalesce: #10608 (though as you point out, we can't quite use coalesce for this).

Reply:
A quick check with choose shows that it's only about 2/3 as fast:

```
CaseWhenBench32/1048576/0              48183534 ns     48183323 ns           13 bytes_per_second=334.659M/s
CaseWhenBench64/1048576/0              48610718 ns     48610432 ns           13 bytes_per_second=660.866M/s
CaseWhenBench32/1048576/99             48226819 ns     48226677 ns           15 bytes_per_second=334.327M/s
CaseWhenBench64/1048576/99             48891957 ns     48892172 ns           13 bytes_per_second=656.996M/s
```

though note we have to call both first_true_in, then fill_null, to get the 'else' behavior which doesn't help.